pkg/gardenlet/controller/managedseed/add.go

```diff
@@ -0,0 +1,192 @@
+			Reconciler:              r,
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+			RateLimiter:             r.RateLimiter,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	seedName := confighelper.SeedNameFromSeedConfig(cfg.SeedConfig)
+
+	if err := c.Watch(
+		source.NewKindWithCache(&seedmanagementv1alpha1.ManagedSeed{}, gardenCluster.GetCache()),
+		controllerutils.EnqueueWithJitterDelay(r.Config),
+		r.ManagedSeedFilterPredicate(seedName),
+		&predicate.GenerationChangedPredicate{},
+	); err != nil {
+		return err
+	}
+
+	return c.Watch(
+		source.NewKindWithCache(&gardencorev1beta1.Seed{}, gardenCluster.GetCache()),
+		mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapSeedToManagedSeed), mapper.UpdateWithNew, c.GetLogger()),
+		r.ManagedSeedFilterPredicate(seedName),
+	)
+}
+
+// ManagedSeedFilterPredicate returns the predicate for ManagedSeed and Seed events.
+func (r *Reconciler) ManagedSeedFilterPredicate(seedName string) predicate.Predicate {
+	return &managedSeedFilterPredicate{
+		seedName: seedName,
+		reader:   r.GardenClient,
+	}
+}
+
+type managedSeedFilterPredicate struct {
+	ctx      context.Context
+	reader   client.Reader
+	seedName string
+}
+
+func (p *managedSeedFilterPredicate) InjectStopChannel(stopChan <-chan struct{}) error {
+	p.ctx = contextutil.FromStopChannel(stopChan)
+	return nil
+}
+
+func (p *managedSeedFilterPredicate) Create(e event.CreateEvent) bool {
+	return p.filterManagedSeed(e.Object)
+}
+
+func (p *managedSeedFilterPredicate) Update(e event.UpdateEvent) bool {
+	return p.filterManagedSeed(e.ObjectNew)
+}
+
+func (p *managedSeedFilterPredicate) Delete(e event.DeleteEvent) bool {
+	return p.filterManagedSeed(e.Object)
+}
+
+func (p *managedSeedFilterPredicate) Generic(_ event.GenericEvent) bool { return false }
+
+// filterManagedSeed checks if the ManagedSeed references a Shoot scheduled on a Seed, for which the gardenlet is responsible.
+func (p *managedSeedFilterPredicate) filterManagedSeed(obj client.Object) bool {
+	var managedSeed *seedmanagementv1alpha1.ManagedSeed
+	switch obj := obj.(type) {
+	case *seedmanagementv1alpha1.ManagedSeed:
+		managedSeed = obj
+	case *gardencorev1beta1.Seed:
+		managedSeed = &seedmanagementv1alpha1.ManagedSeed{}
+		if err := p.reader.Get(p.ctx, kutil.Key(gardencorev1beta1constants.GardenNamespace, obj.Name), managedSeed); err != nil {
+			return false
+		}
+	default:
+		return false
+	}
+
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		return false
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := p.reader.Get(p.ctx, kutil.Key(managedSeed.Namespace, managedSeed.Spec.Shoot.Name), shoot); err != nil {
+		return false
+	}
+
+	if shoot.Spec.SeedName == nil {
+		return false
+	}
+
+	if shoot.Status.SeedName == nil || *shoot.Spec.SeedName == *shoot.Status.SeedName {
+		return *shoot.Spec.SeedName == p.seedName
+	}
+
+	return *shoot.Status.SeedName == p.seedName
+}
```

Comment:
This function is quite complex. Can you split it and create dedicated functions for the various input types for `obj`?